### ISSUE TITLE
Feature: Home Assistant Add-on

### ISSUE THREAD

**Issue by CpuID** [longzheng/open-dynamic-export#63]
Does it make sense to package this project as a Home Assistant Add-on?

Context: a lot of people that intend to run this may already have a hardware device that runs Home Assistant (eg. a small embedded appliance), and Home Assistant Add-ons are mostly just a container with some metadata.

This would make it a streamlined process to install/setup potentially...

https://developers.home-assistant.io/docs/add-ons/

**Comment by longzheng** [comment]
Yes I agree. Unfortunately at the moment the docker compose uses multiple containers (an InfluxDB container), which Home Assistant addons natively support.

@PaulSchulz mentioned he was going to look into this too https://github.com/longzheng/open-dynamic-export/issues/61#issuecomment-2527119857

**Comment by CpuID** [comment]
Sounds good 👍whoever gets time to write up the metadata I guess :) It may also require dropping arm64 Docker images on Docker Hub (in addition to amd64), as a lot of the embedded hardware that Home Assistant recommends are ARM based. That likely just involves adding 2 architectures to https://github.com/longzheng/open-dynamic-export/blob/main/.github/workflows/publish-docker.yml - can be done via a PR at some point.

Adding a link to the Home Assistant InfluxDB add-on for anyone curious: https://github.com/hassio-addons/addon-influxdb

**Comment by longzheng** [comment]
I've given this a bit more thought and I think we can make the InfluxDB completely optional.

Originally I was thinking the system might depend on InfluxDB for data storage/query (eg storing inverter/meter samples to send to CSIP server for dynamic connection telemetry requirements), but currently it's been implemented as an in-memory store and has been fine. I may take it a step further with a SQLite database in the future.

Currently I use InfluxDB for some metric logging which I use to help debug and build a dashboard (that I intended to package with hardware as a plug and play product). If I make InfluxDB an optional config thing, then that "package" can be the one that bundles InfluxDB than this project itself.

A single container will life easier with setup/debugging too.

**Comment by CpuID** [comment]
One nice thing about retaining InfluxDB protocol wise - you can just have an on/off feature toggle to send the metrics still, and if someone wants to use a different metrics store for whatever purpose they have, suggest telegraf as a translation layer (eg. if someone wanted Prometheus as their preferred datastore)

https://github.com/influxdata/telegraf/tree/master/plugins/inputs/influxdb_listener

No point writing native outputs for every metrics store out there :) SQLite is tempting though (very lightweight)

**Comment by PaulSchulz** [comment]
I have managed to write a minimal Dockerfile which successfully runs Open Dynamic Export on Home Assistant. It doesn't start as a HA Add-on yet.. not sure why.

I will commit to by github repo under a 'home-assistant' branch.

**Comment by PaulSchulz** [comment]
https://github.com/PaulSchulz/open-dynamic-export/tree/home-assistant-addon

**Comment by longzheng** [comment]
@PaulSchulz I've just pushed a change that makes InfluxDB optional, so you can even just run the app with Node.js directly. Hopefully that can simplify the HA addon.

**Comment by JimboHamez** [comment]
If this is packaged as a HA add-on will the variables be exposed to HA for use on the Energy page?  I currently use the Sunspec addon to retrieve data from my Fronius Gen24 using Modbus TCP with a polling rate of 5 Secs.  Given the CSIP spec requirement for a polling rate 300ms (United Energy/Powercor) it seems to make sense to use Open Dynamic Export as the source for this data.

**Comment by PaulSchulz** [comment]
The energy page doesn't display the operating envelope that CSIP transmits.
At this stage it would need to be displayed in a separate page.

The power usage/meter information received back from from Modbus on the
inverter (if supported) is output to MQTT which can be read into a HA
sensor (with NodeRED), which can be displayed in the Energy page.

**Comment by JimboHamez** [comment]
@PaulSchulz is the output to MQTT an undocumented feature at this stage?  From what I can tell the publish feature does not forward usage data from the inverter or smart-meter, it is information from the limiter used to control the inverter.

**Comment by longzheng** [comment]
> @PaulSchulz is the output to MQTT an undocumented feature at this stage? From what I can tell the publish feature does not forward usage data from the inverter or smart-meter, it is information from the limiter used to control the inverter.
> 
> update: The inverter and meter Modbus data is available from localhost:3000/sunspec/data. Not tested however HA Scrape looks promising

It's documented here https://opendynamicexport.com/configuration/publish

**Comment by JimboHamez** [comment]
> > @PaulSchulz is the output to MQTT an undocumented feature at this stage? From what I can tell the publish feature does not forward usage data from the inverter or smart-meter, it is information from the limiter used to control the inverter.
> > update: The inverter and meter Modbus data is available from localhost:3000/sunspec/data. Not tested however HA Scrape looks promising
> 
> It's documented here https://opendynamicexport.com/configuration/publish

There seems to be a little confusion here,  my original question was if all of the inverter and smart meter data would be available in home assistant as sensors that could be then used in the Energy dashboard.  From the document you pointed to, the MQTT publish feature per my earlier comment does not appear to contain this data, hence the undocumented feature comment.  Is there another 'topic' other than 'limits' to publish the inverter and smart meter data via MQTT?

**Comment by longzheng** [comment]
@JimboHamez Right sorry I didn't read it closely enough.

Currently there's no MQTT topic to get the inverter/meter data out, but it's available via the `coordinatorStatus` API 

<img width="1625" alt="Image" src="https://github.com/user-attachments/assets/57a888de-578d-4ce6-9139-509a1bde5f29" />

**Comment by CpuID** [comment]
hey @PaulSchulz I'll try test out your branch sometime when I get time on my HA box, it looks fairly straightforward

I think the `.dockerignore` change for `config/` might be controversial/not ideal to make it back into this project, for testing its fine but I think there's a likely alternative path forward for config...

----

re configs, it looks like HA addons can be given `addon_config` ref https://developers.home-assistant.io/docs/add-ons/configuration/
 which can be in the form of extra "files" or whatever else

thinking aloud, how do we feel about:

- `config/config.json` being fed in as a file (ideally from a JSON blob)
- `config/cert.pem` and `config/key.pem` being fed in as files (provided by Long on request)
- `.env` we might need to work out the right way to handle it, there's an `environment` attribute for addons but that would imply a one-size-fits-all set of values, and it would be nice to be able to modify them. either a file or a way to specify the content of them would be ideal

any other files we'd want to configure as addon metadata...? I was thinking that's something I could try work on, getting the syntax together for it all and testing it :)

**Comment by dstu76** [comment]
Thanks @PaulSchulz
